File: LeetCode/169MajorityElement/MajorityElement.py
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        # Boyer-Moore alorithm
        res = nums[0]
        count = 0
        for i in range(len(nums)):
            print(count, res)
            if count == 0 or nums[i] == res:
                res = nums[i]
                count += 1
            else:
                count -= 1
        return res

        # # HashMap method -> Time: O(n), Space: O(n)
        # d = {}
        # n = len(nums)
        # majority = n//2
        # if n%2:
        #     majority += 1
        # for i in range(n):
        #     if nums[i] not in d:
        #         d[nums[i]] = 0
        #     d[nums[i]] += 1
        #     if d[nums[i]] >= majority:
        #         return nums[i]

File: LeetCode/169MajorityElement/MajorityElement.cpp
class Solution {
public:
    int majorityElement(vector<int>& nums) {
        // Boyer-Moore Method -> Time: O(n), Space: O(1)
        int res, count = 0;
        for (int i = 0; i < nums.size(); i++) {
            if (count == 0 || res == nums[i]) {
                res = nums[i];
                count++;
            }
            else {
                count--;
            }
        }
        return res;

        // // HashMap method -> Time: O(n), Space: O(n)
        // unordered_map<int, int> m;
        // int n = nums.size();
        // int majority = n/2;
        // if (n%2) {
        //     majority++;
        // }
        // for (int i = 0; i < n; i++) {
        //     if (m.count(nums[i]) == 0) {
        //         m[nums[i]] = 0;
        //     }
        //     m[nums[i]] += 1;
        //     if (m[nums[i]] >= majority) {
        //         return nums[i];
        //     }
        // }
        // return 0;
    }
};
